src/Core/Services/MetadataProviders/CosmosSqlMetadataProvider.cs

```diff
@@ -590,9 +597,11 @@ public string GetDefaultSchemaName()
             return string.Empty;
         }
 
+        // Use the runtimeConfigProvider to ensure we have current up to date
+        // state on development mode in case a hot
         public bool IsDevelopmentMode()
         {
-            return _runtimeConfig.IsDevelopmentMode();
+            return _runtimeConfigProvider.GetConfig().IsDevelopmentMode();
```

Comment:
if after hot-reload, you're intending to create a new CosmosSqlMetadataProvider, is this change needed? Same with associated function comment.

Reply:
We might not always be creating a new CosmosSqlMetadataProvider, and this just makes sure that no matter what happens outside of this class that we are always returning an answer that is consistent with the current mode

Reply:
That means CosmosSqlMetadataProvider will have inconsistent state: this class could have internal state with old config, and may have *some* helpers return newer config? That doesn't seem right.

You'll need to clarify when new `CosmosSqlMetadataProvider` objects are created and when they are not -> "We might not always be creating a new CosmosSqlMetadataProvider." A better solution here would be to save `_runtimeConfigProvider.GetConfig().IsDevelopmentMode();` as private state to the metadata provider, and update that value when a hotreload event is detected. 

Further analysis shows that having `MetadataProvider` classes expose an `IsDevelopment()` mode helper is... odd design. Please open a refactoring GitHub issue which instructs us to reconsider (and possibly remove) `IsDevelopmentMode()` from the metadataprovider classes. The inclusion of that function seems like a hacky workaround for getting access to "isDevelopmentMode" from code that doesn't have access to the RuntimeConfig -> e.g. SqlQueryStructure classes. Development mode is not set per metadata provider. It should still be fine to have private internal state tracking development mode, so metadata provider logic know whether to emit dev mode error messages or not. But exposing a public getter for a property of the runtime config doesn't make sense.

Reply:
Context from Aaron:
> It is a weird function because all it is doing is replicating the work that the provider is doing. The reason I favored asking the provider instead of saving that state internally is because at some point we might want to avoid having to refresh every single dependency on the runtimeconfig if those portions that we hot-reload are not related to the particular dependency in question. But in that case, this helper function still needs to return a consistent value with the runtimeconfig. Because this helper function doesn't depend on any internal state, but just asks the provider. It should be refactored out imo, as you suggested

This depends on your event notification design: 

- What and how many events will you emit?
- Dependencies can listen for events related to the runtime config properties they care about.

> Because this helper function doesn't depend on any internal state, but just asks the provider. It should be refactored out imo, as you suggested

Please open a refactoring GitHub issue which instructs us to reconsider (and possibly remove) IsDevelopmentMode() from the metadataprovider classes. Add that issue number here before resolving.

Reply:
My reasoning for leaving this was so that we don't add a dependency on the mode to this class with respect to hot-reloading the config when the class itself doesn't depend on that state. This class doesn't use the mode internally and this function is simply passing through what the `RuntimeConfigProvider`'s config says about mode.

So ultimately, if we optimize the hot-reload process to only refresh those dependencies that have changed in the config, it would be beneficial to not have things like this stored internally because the class itself doesn't depend on it, and so now we have added a dependency that we should not have to add in the case of a hot-reload. For now, this shouldn't matter because we will refresh all dependencies to start with.

I will save this state then on construction and we can just return that state, so that the `CosmosMetadataProvider`, but this does mean that this function is not returning the current runtimeconfig's development mode state, but whatever that state was when the `CosmosMetadataProvider` was constructed. And that is a bit weird because the `CosmosMetadataProvider` itself doesn't depend on this state.

will add a task to refactor this out.